### ISSUE TITLE
NCCL stalling between batches

### ISSUE THREAD

**Issue by ehoelzl** [NVIDIA/nccl#329]
I'm having an issue while running NCCL in a distributed multi-node (1 GPU / Node) setting. The stalling happens right after an `all_reduce` call of a quite large tensor (~170MB). 

The communication itself takes under 1 second, but after each reduction, the whole training stalls for 5 seconds before proceeding to the next batch.

I'm using NCCL 2.4.8 with CUDA 10.1 and PyTorch 1.4.1

I've tried modifying some of the environment variables from [here](https://docs.nvidia.com/deeplearning/sdk/nccl-archived/nccl_248/nccl-developer-guide/docs/env.html#nccl-max-nrings), reducing the batch size from 128 to 64 and 32.

Also, I've tried Open-MPI and Gloo backends, which increases communication time, but the stalling does not happen.

I'm suspecting this is cause by cache emptying, since if i add the call `torch.cuda.empty_cache()` between batches, then the stalling disappears (the 5 seconds stall is still there, but taken by `torch.cuda.empty_cache()`. 

Any help regarding this issue would be very much appreciated. Thanks

**Comment by sjeaugey** [comment]
Are you sure all ranks are well synchronized ? That sounds like the case in #146 where on some ranks, some workload is scheduled before NCCL on some ranks and after NCCL on some others. So after the NCCL call completes, some GPU still have a long tail of work to do, which is why cuda.empty_cache() takes a long time to complete.

**Comment by ehoelzl** [comment]
> Are you sure all ranks are well synchronized ? That sounds like the case in #146 where on some ranks, some workload is scheduled before NCCL on some ranks and after NCCL on some others. So after the NCCL call completes, some GPU still have a long tail of work to do, which is why cuda.empty_cache() takes a long time to complete.

Thanks for your reply. I actually did some more benchmarks regarding this issue. Basically, our system runs on Kubernetes, with multiple workers and 1 GPU per worker. 

I ran some tests for `all_reduce` on increasing tensor sizes, testing for the 3 backends (`MPI`, `GLOO` and `NCCL`), both for `float16` and `float32` tensors on GPU and CPU. As you can see from the graphs (below), NCCL is orders of magnitude worse than its counterparts.

Now this can be due to mis-configuration of our Kubernetes cluster (on GCP), which only has `eth0` and `lo` network interfaces, but I'm not 

![fp32vsfp16](https://user-images.githubusercontent.com/17425020/82233570-8231c080-9930-11ea-98dc-204c8900adaa.png)

(`mpihvd` stands for `MPI` used with `horovod`)

**Comment by sjeaugey** [comment]
NCCL 2.4.8 does not have auto tuning for GCP, it was added in 2.5.6. So, you might want to upgrade NCCL, or add NCCL_SOCKET_NTHREADS=4 to your environment.

Also, it might help if you could mention the type of GCP instance you are running, the number of instances you run on, and the time it takes for the NCCL operation to complete at 170 MB (which is the size you mentioned in your original comment). That way we can assess whether you're getting the best performance out of the hardware.

**Comment by ehoelzl** [comment]
@sjeaugey Thanks for your input. Upgrading NCCL to the newest version does improve the performance quite a bit. We now have a similar performance than with OpenMPI.

As to answer your question related to GCP: We are using `n1-standard-4` machines with `Tesla T4` GPUs on a Kubernetes Cluster.

For a 400 MB tensor, we have communication times between 1.8 and 3.3 seconds (for each new cluster we create, times can vary between this range). These changes might be due to proximity of machines.

**Comment by sjeaugey** [comment]
According to this page :
https://cloud.google.com/compute/docs/machine-types#n1_machine_type
`n1-standard-4` should be able to achieve 10Gb/s, which is 1.2 GB/s. So a 400MB allreduce should take 400*2*(nranks-1)/(1200*nranks). Which should be around 0.6s... 1.8s is 3 slower than it should be.

It would be good to run the NCCL tests (https://github.com/nvidia/nccl-tests) on the number of Nodes/GPUs you are interested in, and provide the output. We'd be able to see whether we're close to that speed or whether something is off.

**Comment by ehoelzl** [comment]
@sjeaugey Thank you for being so quick to reply ! I actually figured out the bug that made everything slower than it should be: The code limited the number of CPUs on our machines to 1 (while the machines have 4). Updating NCCL and changing this gives us much better run times (around 0.6s) for such large tensors. 

Thank you for your help, I very much appreciated it.

I'll be closing this issue as it has been resolved.
